### PR TITLE
Feature/blocked rule

### PR DESCRIPTION
**Pull request by x3medima17**
Added a rule that can block a certain amount of assets for a specific user

### CODE REVIEW

File: config/context/application.properties [hunk 1]
```diff
@@ -2,10 +2,10 @@ credential.accountId=brvs@brvs
 credential.pubkeyFilePath=config/context/pub.key
 credential.privkeyFilePath=config/context/priv.key
 brvs.localhostname=localhost
-brvs.userDomains=users
+brvs.userDomains=trading
```
Comment (Mingela) [inline config/context/application.properties]: there is no need to commit this file

File: brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    return balance - amount >= limit;
+
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public boolean isSatisfiedBy(Transaction transaction) {
+    logger.info("New transaction arrived");
```
Comment (Mingela) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java]: I do not think we need such a logger message

File: brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+import java.security.KeyPair;
+import javax.json.Json;
+import javax.json.JsonReader;
+import jp.co.soramitsu.iroha.java.IrohaAPI;
+import jp.co.soramitsu.iroha.java.Query;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class BlockRule implements Rule {
```
Comment (Mingela) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java]: As I understood this is a rule validating the minimal asset balance for the account. Am I correct? If so it should be renamed

File: brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+  private static final Logger logger = LoggerFactory.getLogger(BlockRule.class);
+
+  private IrohaAPI irohaAPI;
+  private String blockerAccount = "blocker@trading";
```
Comment (Mingela) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java]: it should be passed as an constructor argument

File: brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  private String brvsAccountId;
+  private KeyPair keyPair;
+
+  public BlockRule(IrohaAPI irohaAPI, String brvsAccountId,
+      KeyPair keyPair) {
+    this.irohaAPI = irohaAPI;
+    this.brvsAccountId = brvsAccountId;
+    this.keyPair = keyPair;
+
```
Comment (Mingela) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java]: empty line

File: brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+  }
+
```
Comment (Mingela) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java]: empty line

File: brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+  private boolean validateTransfer(Commands.Command transfer, Transaction tx) {
+    String creator = tx.getPayload().getReducedPayload().getCreatorAccountId();
+    String from = transfer.getTransferAsset().getSrcAccountId();
+    logger.info("Creator: {}\nFrom: {}", creator, from);
+    if (!creator.equals(from)) {
+      return true;
```
Comment (Mingela) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java]: did you mean `false`?

File: brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    }
+    String assetId = transfer.getTransferAsset().getAssetId();
+    double amount = Double.parseDouble(transfer.getTransferAsset().getAmount());
+    logger.info("Asset: {}\nAmount: {}", assetId, amount);
```
Comment (Mingela) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java]: I'd really recommend to refactor the logs

File: brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+    String key = assetId.replace("#", "___");
```
Comment (Mingela) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java]: this is not transparent as well, should be passed as an argument I think. Different projects may want to use different separators etc

File: brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    JsonReader reader = Json.createReader(new StringReader(detail));
+    String str_limit = null;
+    try {
+      str_limit = reader
+          .readObject()
+          .getJsonObject(blockerAccount)
+          .getString(key);
+    } catch (Exception ex) {
+      logger.info("Exception on parsing json detail: {}", ex);
+      return true;
```
Comment (Mingela) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java]: `false`?

File: brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+        queryResponse.getAccountAssetsResponse()
+            .getAccountAssetsList()
+            .stream()
+            .filter(asset -> asset.getAssetId().equals(assetId))
+            .mapToDouble(asset -> Double.parseDouble(asset.getBalance()))
+            .sum();
+
+    logger.info("Limit: {}\nBalance: {}", limit, balance);
+    return balance - amount >= limit;
+
```
Comment (Mingela) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java]: empty line

File: config/context/spring-context.xml [hunk 12]
```diff
@@ -106,9 +106,18 @@
...
+    class="iroha.validation.rules.impl.BlockRule">
+    <constructor-arg name="irohaAPI" ref="irohaAPI"/>
+    <constructor-arg name="brvsAccountId" value="${credential.accountId}"/>
+    <constructor-arg name="keyPair" ref="keyPair"/>
+  </bean>
+
   <util:list id="rules" value-type="iroha.validation.rules.Rule">
     <ref bean="sampleRule"/>
-    <ref bean="newBrvsRule"/>
+    <ref bean="blockRule"/>
```
Comment (Mingela) [inline config/context/spring-context.xml]: Do not commit the change of rules list to the general config. Provide a test instead please

**Review by Mingela — CHANGES_REQUESTED** [review]

File: brvs-rules/build.gradle [hunk 13]
```diff
@@ -7,6 +7,10 @@ dependencies {
     compile('org.slf4j:slf4j-api:1.7.25')
     compile('org.slf4j:slf4j-simple:1.7.25')
 
+    compile "javax.json:javax.json-api:1.0"
```
Comment (Mingela) [inline brvs-rules/build.gradle]: I was able to parse details without these dependencies, check `NewBrvsRule`

File: brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java (hunk 7 above)
Comment (x3medima17) [reply brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java]: If this asset is not set in the detail, it means that it is not blocked, so no further check is needed

File: brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java (hunk 10 above)
Comment (x3medima17) [reply brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java]: Same as above, if there is no such key, then any amount is valid

File: brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java (hunk 3 above)
Comment (Alexey-N-Chernyshov) [reply brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java]: Please, add comment describing the logic

File: brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+    logger.info("Asset: {}\nAmount: {}", assetId, amount);
+
+    String key = assetId.replace("#", "___");
+    QueryResponse queryResponse = irohaAPI.query(
+        Query
+            .builder(brvsAccountId, 1L)
+            .getAccountDetail(creator, blockerAccount, key)
+            .buildSigned(keyPair));
+
+    String detail = queryResponse.getAccountDetailResponse().getDetail();
```
Comment (dolgopolovwork) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java]: I'd rather check if the response has errors.

File: brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+
+    JsonReader reader = Json.createReader(new StringReader(detail));
+    String str_limit = null;
+    try {
+      str_limit = reader
+          .readObject()
+          .getJsonObject(blockerAccount)
+          .getString(key);
+    } catch (Exception ex) {
+      logger.info("Exception on parsing json detail: {}", ex);
```
Comment (dolgopolovwork) [inline brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java]: I think it's better to use `error` instead of `info`

Comment (dolgopolovwork) [reply brvs-rules/src/main/java/iroha/validation/rules/impl/BlockRule.java]: What is `{}`?

**Comment by Mingela** [comment]
I think this is a stale one for sure 🔐

**Pull request closed without merging**